<img src="https://raw.githubusercontent.com/EO-College/cubes-and-clouds/main/icons/cnc_3icons_process_circle.svg"
     alt="Cubes & Clouds logo"
     style="float: center; margin-right: 10px;" />

<img src="https://raw.githubusercontent.com/pangeo-data/pangeo.io/refs/heads/main/public/Pangeo-assets/pangeo_logo.png"
     alt="Pangeo logo"
     style="float: center; margin-right: 10px;" />

# 2.4 Data chunking with Pangeo

## Data chunking

<div class="alert alert-info">
<i class="fa-question-circle fa" style="font-size: 22px;color:#666;"></i> <b>Overview</b>
    <br>
    <br>
    <b>Questions</b>
    <ul>
        <li>What is chunking and why does it matter?</li>
        <li>How can we utilize chunking to make our processing more efficient?</li>
    </ul>
    <b>Objectives</b>
    <ul>
        <li>Explore chunking of data</li>
        <li>Learn about the Zarr file format</li>
    </ul>
</div>

## Context

As explained in Section 2.4 - Formats and Performance, when working with large data files or collections, it is often impractical to load the entire dataset into the memory of a single computer at once. This is where the Pangeo ecosystem is particularly useful. In Section 2.3 - Data Access, we discussed the concept of lazy loading. Xarray enables lazy processing of data in __chunks__, meaning the dataset is divided into manageable pieces. By reading and processing the data in these chunks, we can efficiently handle large datasets on a single computer or scale the processing to a distributed computing cluster using Dask (e.g., on the cloud or high-performance computing environments).

How we process these chunks in a parallel environment to scale our computation vertically is discussed in [2.4 dask](./dask.ipynb). In this notebook, you will explore the concept of chunks through various exercises.

Processing data piece by piece is more efficient when both our input and output data are also stored in chunks. As introduced in Section 2.4 - Formats and Performance, [Zarr](https://zarr.readthedocs.io/en/stable/) is a cloud-native data format and serves as the reference library in the Pangeo ecosystem for storing Xarray multi-dimensional datasets in __chunks__.

## Data
We'll begin with the same sample data retrieval method from the Sentinel-2 STAC collection, as described in Exercise 2.3 - Data Access: Lazy Loading with Pangeo.

The analysis will be similar to our previous exercises, but this time, we’ll use a larger spatial extent to demonstrate the scalability.


## Load Libraries

In [1]:
import pystac_client
import geopandas as gpd
from shapely.geometry import mapping
import stackstac
import warnings
import xarray as xr
import numpy as np
import rioxarray as rio
warnings.filterwarnings("ignore")

## Load Sentinel-2 data

In [2]:
aoi = gpd.read_file('../assets/catchment_outline.geojson', crs="EPGS:4326")
aoi_geojson = mapping(aoi.iloc[0].geometry)
URL = "https://earth-search.aws.element84.com/v1"
catalog = pystac_client.Client.open(URL)
items = catalog.search(
    intersects=aoi_geojson,
    collections=["sentinel-2-l2a"],
    datetime="2019-02-01/2019-04-28"
).item_collection()
sentinel2_l2a = stackstac.stack(items)

In [3]:
sentinel2_l2a

<xarray.DataArray 'stackstac-29b639b4cb6a9c5ad2903b8f5d184d30' (time: 101,
                                                                band: 32,
                                                                y: 20982,
                                                                x: 10980)> Size: 6TB
dask.array<fetch_raster_window, shape=(101, 32, 20982, 10980), dtype=float64, chunksize=(1, 1, 1024, 1024), chunktype=numpy.ndarray>
Coordinates: (12/52)
  * time                                     (time) datetime64[ns] 808B 2019-...
    id                                       (time) <U24 10kB 'S2B_32TPT_2019...
  * band                                     (band) <U12 2kB 'aot' ... 'wvp-jp2'
  * x                                        (x) float64 88kB 6e+05 ... 7.098...
  * y                                        (y) float64 168kB 5.3e+06 ... 5....
    s2:degraded_msi_data_percentage          (time) object 808B 0 0 ... 0 0.2268
    ...                                       ...
    title                                    (band) <U31 4kB 'Aerosol optical...
    gsd                                      (band) object 256B None 10 ... None
    common_name                              (band) object 256B None ... None
    center_wavelength                        (band) object 256B None ... None
    full_width_half_max                      (band) object 256B None ... None
    epsg                                     int64 8B 32632
Attributes:
    spec:        RasterSpec(epsg=32632, bounds=(600000.0, 5090220.0, 709800.0...
    crs:         epsg:32632
    transform:   | 10.00, 0.00, 600000.00|\n| 0.00,-10.00, 5300040.00|\n| 0.0...
    resolution:  10.0

## What is a __Chunk__?

If you closely examine the `sentinel2_l2a` dataset, you'll notice that the `xarray.DataArray` is backed by a `dask.array` with a chunk size of `(1, 1, 1024, 1024)`. The full dataset consists of arrays with dimensions `(101, 32, 20982, 10980)`, totaling 746,592 chunks, which amounts to 5.42 TiB of data loaded into the computer's RAM.

You can view the `dask.array` information by clicking the blue-circled icon in the image below.

![Dask.array](../assets/datasize.png)

By clicking the red-circled triangle icon, you'll see detailed information about the `xarray.DataArray`, including its Coordinates, Indexes, and Attributes.

When creating an `Xarray` object using `stackstac`, we can easily convert a STAC collection into a lazily-loaded, chunked `xarray.DataArray` backed by Dask.

The size and shape of the chunks determine the level of parallelization performed by Dask. Therefore, selecting an appropriate chunk size can significantly impact the performance of your computation.

This is where understanding and effectively using chunking becomes crucial.


In our case, for the moment, we used `stackstac` without specifying the 'chunk' explicitly. The dataset is composed of 8 MiB chunks, each containing 1 time step, 1 band, and a resolution of 1024 x 1024 in the x and y directions.

![chunk_original](../assets/chunk_original.png)

If the chunk size is too small, our workflow will be divided into too many tiny pieces, which can lead to excessive communication and increased distribution overhead.

On the other hand, if the chunk size is too large, there may not be enough memory available to handle the workload, causing the workflow to fail.

The optimal chunk size depends on both your computation and the machine you're using.

For example, 8 MiB is relatively small compared to the typical RAM size available. Dask’s default array chunk size, for instance, is 128 MiB.


In [4]:
import dask
dask.config.get('array.chunk-size')

'128MiB'

## Modifying chunks

Let's try to modify our chunk size.

To modify chunks on your existing `xarray.DataArray` we can use the `chunk` method.
We know that we only need 3 bands for computing the snow index example, so we select only `green`,`swir16` and `scl` to simplify our example.

We would like to have each time series separated in each chunk, then keep all band informnation on one chunk, and let dask to compute x and y coordinate's chunk size.

In [5]:
sentinel2_l2a=sentinel2_l2a.sel(
    band=['green','swir16','scl']).chunk(
    chunks={'time': 1, 'band':3, 'x':'auto','y':'auto'})
sentinel2_l2a

<xarray.DataArray 'stackstac-29b639b4cb6a9c5ad2903b8f5d184d30' (time: 101,
                                                                band: 3,
                                                                y: 20982,
                                                                x: 10980)> Size: 558GB
dask.array<rechunk-merge, shape=(101, 3, 20982, 10980), dtype=float64, chunksize=(1, 3, 2048, 2048), chunktype=numpy.ndarray>
Coordinates: (12/52)
  * time                                     (time) datetime64[ns] 808B 2019-...
    id                                       (time) <U24 10kB dask.array<chunksize=(1,), meta=np.ndarray>
  * band                                     (band) <U12 144B 'green' ... 'scl'
  * x                                        (x) float64 88kB 6e+05 ... 7.098...
  * y                                        (y) float64 168kB 5.3e+06 ... 5....
    s2:degraded_msi_data_percentage          (time) object 808B dask.array<chunksize=(1,), meta=np.ndarray>
    ...                                       ...
    title                                    (band) <U31 372B dask.array<chunksize=(3,), meta=np.ndarray>
    gsd                                      (band) object 24B dask.array<chunksize=(3,), meta=np.ndarray>
    common_name                              (band) object 24B dask.array<chunksize=(3,), meta=np.ndarray>
    center_wavelength                        (band) object 24B dask.array<chunksize=(3,), meta=np.ndarray>
    full_width_half_max                      (band) object 24B dask.array<chunksize=(3,), meta=np.ndarray>
    epsg                                     int64 8B 32632
Attributes:
    spec:        RasterSpec(epsg=32632, bounds=(600000.0, 5090220.0, 709800.0...
    crs:         epsg:32632
    transform:   | 10.00, 0.00, 600000.00|\n| 0.00,-10.00, 5300040.00|\n| 0.0...
    resolution:  10.0

If you look into the details of any variable in the representation above, you'll see that each x and y coordinate's chunk is bigger, and we have less chunks (6666 chunks) than the example before.

Note here from the chunk size, the auto option computed that 2048 for y and x as optimal chunk size if we want to keep the chunk size of time and band as 1 and 3 respectively.  


<div class="alert alert-warning">
    <i class="fa-check-circle fa" style="font-size: 22px;color:#666;"></i> <b>Go Further</b>
    <br>
    <br>
    You can try to apply different ways for specifying chunk.
    <ul>
        <li> chunks = -1 -> the entire array will be used as a single chunk
        <li> chunks = {'x':-1, 'y': 1000} -> chunks of entire _x_ dimension, but splitted every 1000 values on _y_ dimension</li>
        <li> chunks = {'x':-1, 'y': 'auto'} -> Xarray relies on Dask to use an ideal size according to the preferred chunk sizes for _y_ dimension</li>
        <li> chunks = { 'x':-1 ,'y':"500MiB" } -> Xarray seeks the size according to a specific memory target expressed in MiB</li>
        <li> chunks = ( 1, 3, 12048,2048) -> Specifying chunk size in the order of dimension. </li>
    </ul>
</div>

## Defining the chunk size at the creation with Xarray

We can define the chunk size when we create the object.  
This is usually done with Xarray using the `chunks` kwarg when opening a file with `xr.open_dataset` or with `xr.open_mfdataset`, if you create Xarray from your local file.  

In our snow index example, we create Xarray from `stackstac`. As `stackstac`'s default 'chunksize' definition is 1024 for the x and y dimensions, we had that chunk size. We can pass the `chunksize` option to `stackstac` and make it larger.



In [6]:
%%time
sentinel2_l2a = stackstac.stack(items
                                ,assets=['green','swir16','scl']
                               ,chunksize=( 1, 3, 2048,2048)
)
sentinel2_l2a

CPU times: user 36.6 ms, sys: 3.23 ms, total: 39.9 ms
Wall time: 38.7 ms


<xarray.DataArray 'stackstac-a8f4a650a3e74cb9f3d3384cb98f4039' (time: 101,
                                                                band: 3,
                                                                y: 20982,
                                                                x: 10980)> Size: 558GB
dask.array<fetch_raster_window, shape=(101, 3, 20982, 10980), dtype=float64, chunksize=(1, 3, 2048, 2048), chunktype=numpy.ndarray>
Coordinates: (12/52)
  * time                                     (time) datetime64[ns] 808B 2019-...
    id                                       (time) <U24 10kB 'S2B_32TPT_2019...
  * band                                     (band) <U6 72B 'green' ... 'scl'
  * x                                        (x) float64 88kB 6e+05 ... 7.098...
  * y                                        (y) float64 168kB 5.3e+06 ... 5....
    s2:degraded_msi_data_percentage          (time) object 808B 0 0 ... 0 0.2268
    ...                                       ...
    title                                    (band) <U30 360B 'Green (band 3)...
    raster:bands                             (band) object 24B None ... [{'no...
    common_name                              (band) object 24B 'green' ... None
    center_wavelength                        (band) object 24B 0.56 1.61 None
    full_width_half_max                      (band) object 24B 0.045 0.143 None
    epsg                                     int64 8B 32632
Attributes:
    spec:        RasterSpec(epsg=32632, bounds=(600000.0, 5090220.0, 709800.0...
    crs:         epsg:32632
    transform:   | 10.00, 0.00, 600000.00|\n| 0.00,-10.00, 5300040.00|\n| 0.0...
    resolution:  10.0

## So, why chunks?

As explained in **Section 2.4 - Formats and Performance**, chunks are mandatory for accessing files or datasets that are larger than a single computer's memory. If all the data has to be accessed, it can be done sequentially (i.e., chunks are processed one after the other).

Moreover, chunks allow for distributed processing and increased speed for your data analysis, as seen in the next section.

### Chunks and files

Xarray chunking capabilities also depend on the underlying input or output file format used. Most modern file formats allow datasets or single files to be stored using chunks. For example, **NetCDF4** uses chunks when storing a file on the disk via HDF5. Any read of data from a NetCDF4 file will load at least one chunk of that file. So when reading one of its chunks as defined in the `open_dataset` call, Xarray will take advantage of native file chunking and won't need to read the entire file.

However, it is important to note that **Xarray chunks and file chunks are not necessarily the same**. It is a good practice to configure Xarray chunks so that they align well with the input file format chunks (ideally, Xarray chunks should contain one or several input file chunks).


## Zarr storage format

This brings us to our next subject: [Zarr](https://zarr.readthedocs.io/en/stable/).

If we can have our original dataset already 'chunked' and accessed in an optimized way according to its actual byte storage on disk, we won't need to load the entire dataset every time. This can greatly optimize our data analysis, even when working with the entire dataset.

Let's convert our intermediate data into Zarr format so that we can learn what it is. We can keep the data as a `DataArray` or convert it into a `Dataset` before storing it.

We start again by loading data using `stackstac`, but this time, we proceed to the next step: clipping the data and computing the snow index. Then, let's try to save those intermediate results in a Zarr file.


## Load data using stackstac (with specific chunk size)

In [7]:
aoi = gpd.read_file("../assets/catchment_outline.geojson", crs="EPGS:4326")
aoi_geojson = mapping(aoi.iloc[0].geometry)
URL = "https://earth-search.aws.element84.com/v1"
catalog = pystac_client.Client.open(URL)
items = catalog.search(
    intersects=aoi_geojson,
    collections=["sentinel-2-l2a"],
    datetime="2019-02-01/2019-04-28"
).item_collection()
ds = stackstac.stack(items
                                ,assets=['green','swir16','scl']
                               ,chunksize=( 1, 3, 1024,1024)
)

## Computing Snow index


In [8]:
green = ds.sel(band='green')
swir = ds.sel(band='swir16')
scl = ds.sel(band='scl')
ndsi = (green - swir) / (green + swir)
snow = xr.where((ndsi > 0.42) & ~np.isnan(ndsi), 1, ndsi)
snowmap = xr.where((snow <= 0.42) & ~np.isnan(snow), 0, snow)
mask = np.logical_not(scl.isin([8, 9, 3])) 
snow_cloud = xr.where(mask, snowmap, 2)

## Clip the data

In [9]:
aoi_utm32 = aoi.to_crs(epsg=32632)
geom_utm32 = aoi_utm32.iloc[0]['geometry']
snow_cloud.rio.write_crs("EPSG:32632", inplace=True)
snow_cloud.rio.set_nodata(np.nan, inplace=True)
snow_cloud = snow_cloud.rio.clip([geom_utm32])

## Save the intermediate result of a few days in Zarr format

In [10]:
snow_cloud_small=snow_cloud.isel(time=slice(0,3))
snow_cloud_small

<xarray.DataArray 'stackstac-5c4b2cf5fd82c014fa9fd227636071b3' (time: 3,
                                                                y: 3341, x: 2630)> Size: 211MB
dask.array<getitem, shape=(3, 3341, 2630), dtype=float64, chunksize=(1, 1024, 1024), chunktype=numpy.ndarray>
Coordinates: (12/52)
  * time                                     (time) datetime64[ns] 24B 2019-0...
    id                                       (time) <U24 288B 'S2B_32TPT_2019...
    band                                     <U6 24B 'scl'
  * x                                        (x) float64 21kB 6.543e+05 ... 6...
  * y                                        (y) float64 27kB 5.202e+06 ... 5...
    s2:degraded_msi_data_percentage          (time) object 24B 0 0 0
    ...                                       ...
    title                                    <U30 120B 'Scene classification ...
    common_name                              object 8B None
    center_wavelength                        object 8B None
    full_width_half_max                      object 8B None
    epsg                                     int64 8B 32632
    spatial_ref                              int64 8B 0
Attributes:
    _FillValue:  nan

### Before saving, we can test other chunk shape.


In [11]:
snow_cloud_small=snow_cloud_small.chunk(chunks = {'x':'auto', 'y': 'auto'}).to_dataset(name='data')
snow_cloud_small

<xarray.Dataset> Size: 211MB
Dimensions:                                  (time: 3, x: 2630, y: 3341)
Coordinates: (12/52)
  * time                                     (time) datetime64[ns] 24B 2019-0...
    id                                       (time) <U24 288B dask.array<chunksize=(3,), meta=np.ndarray>
    band                                     <U6 24B 'scl'
  * x                                        (x) float64 21kB 6.543e+05 ... 6...
  * y                                        (y) float64 27kB 5.202e+06 ... 5...
    s2:degraded_msi_data_percentage          (time) object 24B dask.array<chunksize=(3,), meta=np.ndarray>
    ...                                       ...
    title                                    <U30 120B 'Scene classification ...
    common_name                              object 8B None
    center_wavelength                        object 8B None
    full_width_half_max                      object 8B None
    epsg                                     int64 8B 32632
    spatial_ref                              int64 8B 0
Data variables:
    data                                     (time, y, x) float64 211MB dask.array<chunksize=(1, 3341, 2630), meta=np.ndarray>

### Then clean attribute, and save to zarr

In [12]:
%%time

def remove_attrs(obj, to_remove):
    new = obj.copy()
    new.attrs = {k: v for k, v in obj.attrs.items() if k not in to_remove}
    return new

def encode(obj):
    object_coords = [name for name, coord in obj.coords.items() if coord.dtype.kind == "O"]
    return obj.drop_vars(object_coords).pipe(remove_attrs, ["spec", "transform"])

snow_cloud_small.pipe(encode).to_zarr('test.zarr',mode='w')

CPU times: user 2.7 s, sys: 511 ms, total: 3.21 s
Wall time: 11 s


<div class="alert alert-warning">
    <i class="fa-check-circle fa" style="font-size: 22px;color:#666;"></i> <b>Exercise</b>
    <br>
    <ul>
        <li>What about saving the data in Netcdf format? `ls -la test.zarr` and  `ls -la test.zarr/nobs `</li>
        <li>You can try to explore the zarr file you just created using `ls -la test.zarr` and  `ls -la test.zarr/nobs `</li>
        <li>You can explore zarr metadata file by `cat test.zarr/.zmetadata` </li>
        <li>Did you find the __chunks__ we defined previously in your zarr file? </li>
    </ul>
</div>

## Lets compare how the zarr and netcdf file are stored

In [13]:
xr.open_zarr('test.zarr').to_netcdf('test.nc')

In [14]:
!du -sh test.zarr/ test.nc

1.7M	test.zarr/
202M	test.nc


In [15]:
!ls  -la test.zarr/

total 188
drwxrwxr-x 39 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 .
drwxr-xr-x  5 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 ..
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users     2 Feb 11 10:47 .zattrs
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users    24 Feb 11 10:47 .zgroup
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users 23291 Feb 11 10:47 .zmetadata
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 band
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 constellation
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 created
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 data
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 earthsearch:boa_offset_applied
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 earthsearch:payload_id
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865

In [16]:
!ls  -la test.zarr/data

total 1132
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users   6144 Feb 11 10:47 .
drwxrwxr-x 39 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users   6144 Feb 11 10:47 ..
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users    369 Feb 11 10:47 .zarray
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users    653 Feb 11 10:47 .zattrs
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users 366508 Feb 11 10:47 0.0.0
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users 462125 Feb 11 10:47 1.0.0
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users 309608 Feb 11 10:47 2.0.0


In [17]:
!ls  -la test.zarr/

total 188
drwxrwxr-x 39 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 .
drwxr-xr-x  5 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 ..
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users     2 Feb 11 10:47 .zattrs
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users    24 Feb 11 10:47 .zgroup
-rw-rw-r--  1 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users 23291 Feb 11 10:47 .zmetadata
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 band
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 constellation
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 created
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 data
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 earthsearch:boa_offset_applied
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865a-bf2ccb42267e users  6144 Feb 11 10:47 earthsearch:payload_id
drwxrwxr-x  2 6ad00bd4-9f98-41ae-865

In [18]:
!cat test.zarr/.zmetadata | head -n 30

{
    "metadata": {
        ".zattrs": {},
        ".zgroup": {
            "zarr_format": 2
        },
        "band/.zarray": {
            "chunks": [],
            "compressor": null,
            "dtype": "<U6",
            "fill_value": null,
            "filters": null,
            "order": "C",
            "shape": [],
            "zarr_format": 2
        },
        "band/.zattrs": {
            "_ARRAY_DIMENSIONS": []
        },
        "constellation/.zarray": {
            "chunks": [],
            "compressor": null,
            "dtype": "<U10",
            "fill_value": null,
            "filters": null,
            "order": "C",
            "shape": [],
            "zarr_format": 2
        },
        "constellation/.zattrs": {


Zarr format main characteristics are the following:

- Every chunk of a Zarr dataset is stored as a single file (see x.y files in `ls -al test.zarr/data`)
- Each Data array in a Zarr dataset has a two unique files containing metadata:
  - .zattrs for dataset or dataarray general metadatas
  - .zarray indicating how the dataarray is chunked, and where to find them on disk or other storage.


## Conclusion

Understanding chunking is key to optimizing your data analysis when dealing with large datasets. In this exercise, we learned how to optimize data access time and memory resources by using native file chunks loaded by stackstac and instructing Xarray to modify the chunk. Computing on large datasets can be very slow on a single machine, and to optimize your time we may need to parallelize your computations. This is what you will learn in the next exercise with Dask.